api/views/share.py

```diff
@@ -2,18 +2,66 @@
 from rest_framework.response import Response
 
 from rest_framework_json_api import serializers
+from rest_framework.exceptions import PermissionDenied
 
 from django import http
 from django.conf import settings
+from django.db.models.base import ModelBase
+from django.db.models.manager import Manager
+from django.db.models.query import QuerySet
+from django.http import Http404
 from django.views.decorators.http import require_GET
 from django.views.generic.base import RedirectView
-from django.shortcuts import get_object_or_404
 
 from api.pagination import CursorPagination
 from api import serializers as api_serializers
 
 from share.util import IDObfuscator, InvalidID
-from share.models import Source
+from share.models import Source, AbstractCreativeWork
+
+
+def _get_queryset(klass):
+    """
+    Returns a QuerySet from a Model, Manager, or QuerySet. Supports get_object_or_404_or_403.
+
+    Raises a ValueError if klass is not a Model, Manager, or QuerySet.
+    """
+    if isinstance(klass, QuerySet):
+        return klass
+    elif isinstance(klass, Manager):
+        manager = klass
+    elif isinstance(klass, ModelBase):
+        manager = klass._default_manager
+    else:
+        if isinstance(klass, type):
+            klass__name = klass.__name__
+        else:
+            klass__name = klass.__class__.__name__
+        raise ValueError("Object is of type '%s', but must be a Django Model, "
+                         "Manager, or QuerySet" % klass__name)
+    return manager.all()
+
+
+def get_object_or_404_or_403(klass, *args, **kwargs):
+    """
+    Replaces the get_object_or_404 in django.shortcuts.
+
+    Uses get() to return an object, or raises a Http404 exception if the object
+    does not exist.
+
+    klass may be a Model, Manager, or QuerySet object. All other passed
+    arguments and keyword arguments are used in the get() query.
+
+    Note: Like with get(), an MultipleObjectsReturned will be raised if more than one
+    object is found.
+    """
+    queryset = _get_queryset(klass)
+    try:
+        return queryset.get(*args, **kwargs)
+    except queryset.model.DoesNotExist:
+        if AbstractCreativeWork.objects.filter(id=kwargs['pk']).count() == 1:
```

Comment:
`kwargs` won't always contain `'pk'` (like in `source_icon_view` below). Since you can't assume what filters are passed in, it might be better to rework this so you don't have to.